### DISCUSSION TITLE
UpdateBuffer()

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#214] [Q&A] [answered]
Just a quick question about `DeviceContextD3D12Impl::UpdateBuffer()`. It seems to be allocating a temporary block of memory, copying the source data to it, then writing that to the buffer. Could anyone help me understand why that's necessary, and is there any way to avoid the allocation and additional copy?

I try to avoid doing allocations per frame in my app. Really hoping there is some way to structure the original source data so that it is already ready to be copied directly?

**Comment by TheMostDiligent** [comment, marked as answer]
> Could anyone help me understand why that's necessary, and is there any way to avoid the allocation and additional copy?

It is because GPU-local memory is generally not directly accessible by GPU. You have to use temporary memory that CPU can write to and GPU can read from.

> is there any way to avoid the allocation and additional copy?

Use dynamic buffers. They allow writing to GPU-readable memory directly.

> I try to avoid doing allocations per frame in my app.

Allocation for Update buffer is very fast. The performance bottleneck there is state transition required to perform the copy operation.

**Reply by RobertMtx** [reply]
I didn't realize the allocation was being made for a special type of memory. I assumed it was allocating standard CPU memory.

I do use dynamic buffers for most of my bulk shader inputs, but I'm considering using `UpdateBuffer()` for a `ConstantBuffer<>` variable. The structure is not large, contains data that will be accessed by every type of shader, and is only updated once per frame. So it seemed like a good fit. From some material I've read, it sounds like constant buffers are often quicker to access on the GPU side. Not sure if that is a specific graphics card trait, or very common, but it seemed like a decent tradeoff.

But yeah, if this is just how it needs to happen for a non-dynamic buffer, then I'm good with it. I thought maybe it was allocating temporary CPU memory just to align it on 16 byte boundaries, or something along those lines, where the original source data could be pre-formatted that way to avoid it.

Thanks for the information!

**Reply by TheMostDiligent** [reply]
> I'm considering using UpdateBuffer() for a ConstantBuffer<> variable. The structure is not large, contains data that will be accessed by every type of shader, and is only updated once per frame.

Yes, if it is once per frame and contains data accessed by all shaders, this totally makes sense.

> I thought maybe it was allocating temporary CPU memory just to align it on 16 byte boundaries

No, this must be allocated from a special pool of memory with GPU read access. GPU can't arbitrary read CPU memory.
